Yesod/Auth/OAuth2/Bitbucket.hs

```diff
@@ -106,32 +105,34 @@ oauth2BitbucketScoped :: YesodAuth m
 oauth2BitbucketScoped clientId clientSecret scopes = authOAuth2 "bitbucket" oauth fetchBitbucketProfile
   where
     oauth = OAuth2
-        { oauthClientId = encodeUtf8 clientId
-        , oauthClientSecret = encodeUtf8 clientSecret
-        , oauthOAuthorizeEndpoint = encodeUtf8 $ "https://bitbucket.com/site/oauth2/authorize?scope=" <> T.intercalate "," scopes
+        { oauthClientId = clientId
+        , oauthClientSecret = clientSecret
+        , oauthOAuthorizeEndpoint = "https://bitbucket.com/site/oauth2/authorize" `withQuery`
+            [ ("scope", encodeUtf8 $ T.intercalate "," scopes)
+            ]
         , oauthAccessTokenEndpoint = "https://bitbucket.com/site/oauth2/access_token"
         , oauthCallback = Nothing
         }
 
-fetchBitbucketProfile :: Manager -> AccessToken -> IO (Creds m)
+fetchBitbucketProfile :: Manager -> OAuth2Token -> IO (Creds m)
 fetchBitbucketProfile manager token = do
-    userResult <- authGetJSON manager token "https://api.bitbucket.com/2.0/user"
-    mailResult <- authGetJSON manager token "https://api.bitbucket.com/2.0/user/emails"
+    userResult <- authGetJSON manager (accessToken token) "https://api.bitbucket.com/2.0/user"
+    mailResult <- authGetJSON manager (accessToken token) "https://api.bitbucket.com/2.0/user/emails"
 
     case (userResult, mailResult) of
         (Right user, Right mails) -> return $ toCreds user (bitbucketEmails mails) token
-        (Left err, _) -> throwIO $ InvalidProfileResponse "bitbucket" err
-        (_, Left err) -> throwIO $ InvalidProfileResponse "bitbucket" err
+        (Left err, _) -> throwIO $ invalidProfileResponse "bitbucket" err
```

Comment:
is there anything with `Alternative` that might help here?

Reply:
Not AFAICT, not without moving up to something like `ExceptT`:

```hs
-- Not tested, but I think this is how you'd do it
result <- runExcepT $ do
    user <- ExceptT $ authGetJSON manager (accessToken token) "https://api.bitbucket.com/2.0/user"
    mails <- ExceptT $ authGetJSON manager (accessToken token) "https://api.bitbucket.com/2.0/user/emails"
    return $ toCreds user (bitbucketEmails mails) token

either (throwIO . invalidProfileResponse "bitbucket") return result
```

With a few ancillary functions, it even makes for a nice `Applicative`:

```hs

throwLeft $ runExcepT $ toCreds' token
    <$> ExceptT (authGetJSON manager (accessToken token) "https://api.bitbucket.com/2.0/user")
    <*> ExceptT (authGetJSON manager (accessToken token) "https://api.bitbucket.com/2.0/user/emails")
  where
    toCreds' t u ms = toCreds u (bitbucketEmails ms) t
    throwLeft = either (throwIO . invalidProfileResponse "bitbucket") return
```

I like this pattern when it's pervasive, but I'm not sure about doing it in a random spot.